src/optimizer/rule/MergeGetNbrsAndDedupRule.cpp

```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/MergeGetNbrsAndDedupRule.h"
+
+#include "optimizer/OptGroup.h"
+#include "planner/PlanNode.h"
+#include "planner/Query.h"
+
+using nebula::graph::Dedup;
+using nebula::graph::GetNeighbors;
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> MergeGetNbrsAndDedupRule::kInstance =
+    std::unique_ptr<MergeGetNbrsAndDedupRule>(new MergeGetNbrsAndDedupRule());
+
+MergeGetNbrsAndDedupRule::MergeGetNbrsAndDedupRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern &MergeGetNbrsAndDedupRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kGetNeighbors,
+                                             {Pattern::create(graph::PlanNode::Kind::kDedup)});
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> MergeGetNbrsAndDedupRule::transform(
+    QueryContext *qctx,
+    const MatchedResult &matched) const {
+    const OptGroupNode *optGN = matched.node;
+    const OptGroupNode *optDedup = matched.dependencies.back().node;
+    DCHECK_EQ(optGN->node()->kind(), PlanNode::Kind::kGetNeighbors);
+    DCHECK_EQ(optDedup->node()->kind(), PlanNode::Kind::kDedup);
+    auto gn = static_cast<const GetNeighbors *>(optGN->node());
+    auto dedup = static_cast<const Dedup *>(optDedup->node());
+    auto newGN = gn->clone(qctx);
+    if (!newGN->dedup()) {
+        newGN->setDedup();
+    }
+    newGN->setInputVar(dedup->inputVar());
+    auto newOptGV = OptGroupNode::create(qctx, newGN, optGN->group());
+    for (auto dep : optDedup->dependencies()) {
```

Comment:
In fact I think this dedup node is still necessary (e.g. request to multiple hosts that the push-down dedup option only dedup result from one host).

Reply:
The removed dedup node is the input of GetNeighbors, rather than dedup the result of GetNeighbors. The execution plan is like:

```text
...-> Dedup -> (vids) -> GetNeighbors -> ...

=>

...-> GetNeighbors -> ...  # dedup vids inside GetNeighbors
```

The benefits of the rule is avoid the redundant src expression evaluation.

Reply:
ok.